src/app/controllers/sap.php

```diff
@@ -3,7 +3,22 @@
 class sap extends Controller {
 
     function index() {
+        $this->load_library('token_generator_lib', 'token_generator');
+        $csrf_token = $this->token_generator->getCSRFToken();
+
+        session_start();
+        if (isset($_SESSION['csrf_token'])) {
```

Comment:
Probably. I don't know what. I don't usually php, and when I do, I copy paste from somewhere nice.